src/fugw/solvers/dense.py

```diff
@@ -128,14 +144,14 @@ def fugw_loss(
         loss_regularization = torch.zeros(1)
         loss = 0
 
-        if alpha != 1 and D is not None:
-            loss_wasserstein = ((D * pi).sum() + (D * gamma).sum()) / 2
+        if alpha != 1 and K1 is not None and K2 is not None:
+            loss_wasserstein = ((K1 @ K2.T) * (pi + gamma)).sum() / 2
```

Comment:
This is a different formula, right ? Any explanation on the change ?

Reply:
$D \simeq K_1 K_2 ^T$ so this is just a factored expression to be consistent with
```
loss_wasserstein = csr_sum(elementwise_prod_fact_sparse(K1, K2, pi + gamma)) / 2
```
which appears in the sparse solver.